pkg/operator/configobservation/auth/external_oidc.go

```diff
@@ -0,0 +1,159 @@
+package auth
+
+import (
+	"encoding/json"
+	"fmt"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/api/features"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/configobservation"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/configobserver/featuregates"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resourcesynccontroller"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apiserver/pkg/apis/apiserver"
+	"k8s.io/klog/v2"
+)
+
+const (
+	apiServerArgumentsPath = "apiServerArguments"
+	argAuthConfig          = "authentication-config"
+
+	SourceAuthConfigCMNamespace = "openshift-config-managed"
+	AuthConfigCMName            = "auth-config"
+	staticAuthConfigPath        = "/etc/kubernetes/static-pod-resources/configmaps/" + AuthConfigCMName + "/auth-config.json"
+)
+
+func NewObserveExternalOIDC(featureGateAccessor featuregates.FeatureGateAccess) configobserver.ObserveConfigFunc {
+	return (&externalOIDC{
+		featureGateAccessor: featureGateAccessor,
+	}).ObserveExternalOIDC
+}
+
+type externalOIDC struct {
+	featureGateAccessor featuregates.FeatureGateAccess
+}
+
+// ObserveExternalOIDC observes the authentication.config/cluster resource
+// and if the type field is set to OIDC, it configures an external OIDC provider
+// to the KAS pods by setting the --authentication-config apiserver argument. It also
+// takes care of synchronizing the structured auth config file into the apiserver's namespace
+// so that it gets mounted as a static file on each node.
+func (o *externalOIDC) ObserveExternalOIDC(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	if o.featureGateAccessor == nil || !o.featureGateAccessor.AreInitialFeatureGatesObserved() {
+		// if we haven't observed featuregates yet, return the existing
+		return existingConfig, nil
+	}
+
+	featureGates, err := o.featureGateAccessor.CurrentFeatureGates()
+	if err != nil {
+		return existingConfig, []error{err}
+	}
+
+	if !featureGates.Enabled(features.FeatureGateExternalOIDC) {
+		return existingConfig, nil
+	}
+
+	listers := genericListers.(configobservation.Listers)
+	auth, err := listers.AuthConfigLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		klog.Warningf("authentications.config.openshift.io/cluster: not found")
```

Comment:
I followed the example of auth_metadata and webhook_authenticator config observers that don't record events for that case; I'll add events to all three of them, since that's what we usually do.